### PR TITLE
Varvara/feature/push notification

### PR DESCRIPTION
**Pull request by varvarashcherbyna7**

### CODE REVIEW

File: app/commands/__init__.py [hunk 1]
```diff
@@ -24,3 +24,19 @@ def db_populate(count: int):
...
             print(f"DB populated by {count} instancies")
+
+    @app.cli.command("create-admin")
+    def create_admin():
+        """Create super admin account"""
+        query = m.SuperUser.select().where(
+            m.SuperUser.email == app.config["ADMIN_EMAIL"]
+        )
+        if db.session.execute(query).first():
+            print(f"User with e-mail: [{app.config['ADMIN_EMAIL']}] already exists")
```
Comment (Nazar-kob) [inline app/commands/__init__.py]: we have with command on back

File: app/controllers/notification.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    push_handler.send_notification(
+        s.PushNotificationMessage(
+            device_tokens=devices,
+            payload=get_notification_payload(
+                notification_type=s.NotificationType.CASE_CREATED.value, case=case
+            ),
+        )
+    )
+
+    log(log.INFO, "[%d] notifications sended", len(devices))
```
Comment (Nazar-kob) [inline app/controllers/notification.py]: why this func is out of class ?

File: app/controllers/notification.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def get_notification_payload(notification_type: s.NotificationType, case: m.Case):
+    return s.PushNotificationPayload(
+        notification_type=notification_type,
+        title=case.title,
+        sub_title=case.sub_title,
+        project_link=case.project_link,
+    )
```
Comment (Nazar-kob) [inline app/controllers/notification.py]: same things as above

File: app/schema/case.py [hunk 4]
```diff
@@ -14,8 +14,8 @@ class CaseOut(BaseModel):
     stacks: list[str]
     screenshots: list[str]
 
-    main_image: str
-    preview_image: str
+    main_image: str | None
+    preview_image: str | None
```
Comment (Nazar-kob) [inline app/schema/case.py]: the schema was changed by the last pull request

File: app/controllers/notification.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    def __init__(self):
+        if PushHandler._is_initialized:
+            log(log.INFO, "Firebase was initialized")
+            return
+
+        cred = credentials.Certificate(
+            current_app.config["GOOGLE_SERVICE_ACCOUNT_PATH"]
+        )
+        firebase_admin.initialize_app(cred)
+        PushHandler._is_initialized = True
```
Comment (Nazar-kob) [inline app/controllers/notification.py:43]: can we make it a singleton?

File: app/controllers/notification.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            )
+
+        except (ValueError, TypeError):
+            log(log.ERROR, "Message arguments invalid")
+            raise HTTPException(code=409, description="Message arguments invalid")
+
+
+def case_created_notify(case: s.PushNotificationPayload):
+    db.session.refresh(case)
+    all_devices: list[m.Device] = db.session.query(m.Device).all()
```
Comment (Nazar-kob) [inline app/controllers/notification.py:75]: use sqlalchemy 2.0

File: app/controllers/notification.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+    devices: list[str] = list()
+
+    if all_devices:
+        for device in all_devices:
+            devices.append(device.token)
+
```
Comment (Nazar-kob) [inline app/controllers/notification.py:82]: we can get it by sqlalchemy 2.0

File: app/controllers/notification.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                        mutable_content=True,
+                    ),
+                    headers={"apns-priority": "5"},
+                ),
+            ),
+        )
+
+        try:
+            messaging.send_multicast(multicast_message=message)
+            log(log.INFO, "Notification sended")
```
Comment (Nazar-kob) [inline app/controllers/notification.py:59]: do the message have return?

File: migrations/env.py
Comment (Nazar-kob) [inline migrations/env.py:1]: not need migrations in admin, backend responses for it

File: migrations/alembic.ini
Comment (Nazar-kob) [inline migrations/alembic.ini:1]: not need migrations in admin, backend responses for it

File: migrations/script.py.mako
Comment (Nazar-kob) [inline migrations/script.py.mako:1]: not need migrations in admin, backend responses for it

File: migrations/versions/b581353d04f2_.py
Comment (Nazar-kob) [inline migrations/versions/b581353d04f2_.py:1]: not need migrations in admin, backend responses for it

**Review by Nazar-kob — CHANGES_REQUESTED** [review]

File: app/controllers/__init__.py [hunk 9]
```diff
@@ -1,4 +1,5 @@
 # flake8: noqa F401
 from .pagination import create_pagination
 from .s3_bucket import S3Bucket
+from .notification import PushHandler, case_created_notify
```
Comment (nixissmtp) [inline app/controllers/__init__.py]: notify_case_created

File: app/controllers/notification.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+from firebase_admin.exceptions import FirebaseError
+from werkzeug.exceptions import HTTPException
+
+from app.common import models as m
+from app import schema as s
+from app.database import db
+from app.logger import log
+
+
+class PushHandler:
```
Comment (nixissmtp) [inline app/controllers/notification.py:31]: need refactor

File: app/controllers/notification.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+            return
+
+        cred = credentials.Certificate(
+            current_app.config["GOOGLE_SERVICE_ACCOUNT_PATH"]
+        )
+        firebase_admin.initialize_app(cred)
+        PushHandler._is_initialized = True
+
+    def send_notification(self, message_data: s.PushNotificationMessage):
+        if not message_data.device_tokens:
```
Comment (nixissmtp) [inline app/controllers/notification.py:46]: make PushNotificationMessage.device_tokens required with validation

File: app/controllers/notification.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+            log(log.INFO, "User has no tokens to push")
+            return
+
+        message = messaging.MulticastMessage(
+            tokens=message_data.device_tokens,
+            data=message_data.payload.dict(),
+            android=messaging.AndroidConfig(
```
Comment (nixissmtp) [inline app/controllers/notification.py]: make android config constant

File: app/controllers/notification.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+                ttl=3600,
+                priority="high",
+            ),
+            apns=messaging.APNSConfig(
```
Comment (nixissmtp) [inline app/controllers/notification.py]: same as android config

File: app/controllers/notification.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+            return {"status": "should be done"}
+
+        except FirebaseError:
+            log(log.ERROR, "Error while sending message")
+            raise HTTPException(
+                code=409,
+                description="Error while sending message",
+            )
+
+        except (ValueError, TypeError):
```
Comment (nixissmtp) [inline app/controllers/notification.py:68]: validation to pydantic

File: app/controllers/notification.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+            log(log.ERROR, "Message arguments invalid")
+            raise HTTPException(code=409, description="Message arguments invalid")
+
+
+def case_created_notify(case: s.PushNotificationPayload):
```
Comment (nixissmtp) [inline app/controllers/notification.py]: notify_case_created

File: app/controllers/notification.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    db.session.refresh(case)
+    all_devices: list[m.Device] = db.session.query(m.Device).all()
+
+    devices: list[str] = list()
```
Comment (nixissmtp) [inline app/controllers/notification.py:77]: devices = [device.token for device in all_devices]

File: app/controllers/notification.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+    if all_devices:
+        for device in all_devices:
+            devices.append(device.token)
+
+    push_handler = PushHandler()
+    push_handler.send_notification(
```
Comment (nixissmtp) [inline app/controllers/notification.py:84]: use pydantic

File: app/controllers/notification.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+        s.PushNotificationMessage(
+            device_tokens=devices,
+            payload=get_notification_payload(
```
Comment (nixissmtp) [inline app/controllers/notification.py:87]: need refactor

File: app/schema/case.py (hunk 4 above)
Comment (nixissmtp) [reply app/schema/case.py]: return only str

File: app/schema/push_notification.py [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+class PushNotificationPayload(BaseModel):
+    notification_type: str
+    title: str
+    sub_title: str
+    project_link: str | None
+
+
+class PushNotificationMessage(BaseModel):
+    payload: PushNotificationPayload
+    device_tokens: list[str] | None
```
Comment (nixissmtp) [inline app/schema/push_notification.py:13]: remove None

File: app/views/case.py [hunk 20]
```diff
@@ -9,6 +9,7 @@
 from app.common import models as m
 from app import schema as s
 from app import forms as f
+from app.controllers import case_created_notify
 from app.logger import log
 from app.database import db
 from app import s3bucket
```
Comment (nixissmtp) [inline app/views/case.py:17]: sqlalchemy
is_(False)

File: app/views/case.py [hunk 21]
```diff
@@ -161,7 +162,10 @@ def create():
             session.add(new_stack)
         session.commit()
 
+        case_created_notify(new_case)
+        log(log.INFO, "Case created. Case: [%s]", new_case)
         flash("Case added!", "success")
+
     if form.errors:
```
Comment (nixissmtp) [inline app/views/case.py:170]: move to up

**Review by nixissmtp — CHANGES_REQUESTED** [review]

**Pull request merged**